fuel/utils.py

```diff
@@ -14,6 +16,347 @@
     buffer_ = buffer  # noqa
 
 
+class Subset(object):
+    """A description of a subset of examples.
+
+    Parameters
+    ----------
+    list_or_slice : :class:`list` or :class:`slice`
+        List of positive integer indices or slice that describes which
+        examples are part of the subset.
+    original_num_examples: int
+        Number of examples in the dataset this subset belongs to.
+
+    Attributes
+    ----------
+    is_list : bool
+        Whether the Subset is a list-based subset (as opposed to a
+        slice-based subset).
+    num_examples : int
+        Number of examples the Subset spans.
+    original_num_examples : int
+        Number of examples in the dataset this subset is part of.
+
+    """
+    def __init__(self, list_or_slice, original_num_examples):
+        self._subset_sanity_check(list_or_slice, original_num_examples)
+        if self._is_list(list_or_slice):
+            list_or_slice = self._beautify_list(list_or_slice)
+        self.list_or_slice = list_or_slice
+        self.original_num_examples = original_num_examples
+
+    def __add__(self, other):
+        if self.original_num_examples != other.original_num_examples:
+            raise ValueError("trying to add two Subset instances with "
+                             "different numbers of original examples, they "
+                             "can't possibly belong to the same dataset")
+        if self.is_list:
+            if other.is_list:
+                return self.__class__(self.list_or_slice + other.list_or_slice,
+                                      self.original_num_examples)
+            else:
+                return self.__class__(self.list_or_slice +
```

Comment:
To reduce copy-paste this branch could make a similar call with swapped arguments.